Added cluster and entry iterators

DrDeano opened the pull request:
This adds iterators to loop over the FAT cluster chain and loop over entries in the directory structure of FAT32.
The tests use hand crafted buffers for a FAT32 filesystem. These are too small for real FAT32 but will still demonstrate the same functionality as if there were real FAT32 buffers.

Part of #216

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -652,6 +659,445 @@ pub fn Fat32FS(comptime StreamType: type) type {
+            pub fn init(allocator: *Allocator, fat_config: FATConfig, cluster: u32, stream: StreamType) (Allocator.Error || Fat32Self.Error || ReadError || SeekError)!EntryIteratorSelf {
+                var cluster_block = try allocator.alloc(u8, fat_config.bytes_per_sector * fat_config.sectors_per_cluster);
+                errdefer allocator.free(cluster_block);
+                var it = try ClusterChainIterator.init(allocator, fat_config, cluster, stream);
```
Does this need an errdefer to clean up?

DrDeano replied:
it does, and forgot to add tests for this init function

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -652,6 +659,445 @@ pub fn Fat32FS(comptime StreamType: type) type {
+                        var long_entry_count = self.cluster_block[self.index] & ~@as(u32, 0x40);
+                        // Allocate a buffer for the long name. 13 for the 13 characters in the long entry
+                        // * 3 as long names are u16 chars so when converted to u8, there could be 3 u8 to encode a u16 (UTF16 -> UTF8)
+                        long_name = try self.allocator.alloc(u8, 13 * 3 * long_entry_count);
```
errdefer to clean up?

DrDeano replied:
The errdefer is higher up (line 948)

SamTebbs33 replied:
Yes, my mistake :)

SamTebbs33 reviewed (commented):
Looking good.

SamTebbs33 reviewed (approved):
:rocket:

DrDeano committed:
Added cluster and entry iterators

This adds iterators to loop over the FAT cluster chain and loop over entries in the directory structure of FAT32.
The tests use hand crafted buffers for a FAT32 filesystem. These are too small for real FAT32 but will still demonstrate the same functionality as if there were real FAT32 buffers.

Added tests for EntryIterator init

Fixed memory leaks
